code_generation/optimizer.py

```diff
@@ -0,0 +1,292 @@
+            if producer.get_outputs("global") != None:
+                outputs.extend(
+                    [
+                        quantity
+                        for quantity in producer.get_outputs("global")
+                        if not isinstance(quantity, NanoAODQuantity)
+                    ]
+                )
+        return outputs
+
+    """
+    Function used to relocate all filters to the top of the ordering
+
+    Args:
+        None
+    Returns:
+        None
+    """
+
+    def MoveFiltersUp(self):
+        new_ordering = []
+        for producer in self.ordering:
+            if isinstance(producer, Filter):
+                new_ordering.insert(0, producer)
+            else:
+                new_ordering.append(producer)
+        for i, prod in enumerate(self.ordering):
+            log.debug(" --> {}. : {}".format(i, prod))
+        for i, prod in enumerate(new_ordering):
+            log.debug(" --> {}. : {}".format(i, prod))
+        self.ordering = new_ordering
+
+    """
+    The main function of this class. During the optimization, a
+    finding a correct ordering is attempted. This is done as follows:
+
+    1. Bring all filters to the beginning of the ordering
+    2. Check if the ordering is already correct. The ordering is correct,
+        if, for all producers in the ordering, all inputs can be found in
+        the outputs of preceeding producers. If the scope is not global,
+        all outputs from producers in the global scope are also considered.
+    3. If the ordering is correct, return.
+    4. If the ordering is not correct,
+        4.1. find all inputs, that have to be produced before the wrong producer
+        4.2. put one producer, which is responsible for creating the input, in front of the wrong producer
+        4.3. repeat from step 2
+
+    The sorting algorithm should take at most 2*(number of producers) steps.
+    If this limit is reached, the optimization is
+    considered to be failed and an Exception is raised.
+    If a missing input cant be found in all outputs,
+    the Optimize function will raise an Exception.
+    """
+
+    def Optimize(self):
+        # first bring filters to the top
+        self.MoveFiltersUp()
+        # run optimization in a while loop
+        counter = 0
+        while not self.optimized:
+            if counter > 2 * self.size + 1:
+                log.error("Could not optimize ordering")
+                log.error("Please check, if all needed producers are activated")
+                raise Exception
+            wrongProducer, wrong_inputs = self.check_ordering()
+            producers_to_relocate = self.find_inputs(wrongProducer, wrong_inputs)
+            if len(producers_to_relocate) == 0:
+                self.optimized = True
+            else:
+                for producer_to_relocate in producers_to_relocate:
+                    counter += 1
+                    producerIndex = self.get_position(wrongProducer)
```

Comment:
better use this function in place below as done for producer_to_relocate

Reply:
true !